### ISSUE TITLE
Cloudflare Zone not found.

### ISSUE THREAD

**Issue by dprus** [troglobit/inadyn#275]
Hi,

I'm trying to use the Cloudflare plug-in but I seem to be having issues.

I generated an API token at cloudflare.com. The token as edit permissions (Zone.DNS) on all zones. However, after running inadyn I get an error (I've redacted personally identifiable information):

> inadyn[20]: Checking IPv4 address x.x.x.x ...                                                                                                                 
> inadyn[20]: IPv4 address x.x.x.x is valid.                                                                                                                    
> inadyn[20]: Current IP# x.x.x.x at default@cloudflare.com                                                                                                     
> inadyn[20]: Update forced for alias <alias>, new IP# x.x.x.x                                                                                  
> inadyn[20]: User: yyy@zzz.aaa Zone: zzz.aaa                                                                                                              
> inadyn[20]: Id query, initiating HTTPS ...                                                                                                                         
> inadyn[20]: Id query, connecting to api.cloudflare.com([104.19.193.29]:443)                                                                                        
> inadyn[20]: Certificate OK                                                                                                                                         
> inadyn[20]: SSL server cert subject: C=US,ST=CA,L=San Francisco,O=CloudFlare\, Inc.,CN=api.cloudflare.com                                                          
> inadyn[20]: SSL server cert issuer: C=US,ST=CA,L=San Francisco,O=CloudFlare\, Inc.,CN=CloudFlare Inc ECC CA-2                                                      
> inadyn[20]: Successfully sent HTTPS request!                                                                                                                       
> inadyn[20]: Successfully received HTTPS response (1020 bytes)!                                                                                                     
> inadyn[20]: Cloudflare says our request was invalid.                                                                                                               
> inadyn[20]: Zone 'zzz.aaa' not found.   

All the redacted information is correct. It's displaying my correct e-mail address, the correct hostname to be updated, and the correct name of the zone.

I'm a bit stuck now so if someone could help me out I'd appreciate it. Thanks.

**Comment by troglobit** [comment]
Never tried Cloudflare myself, but maybe @SimonPilkington can help?  He's the main contributor of the Cloudlflare plugin in Inadyn.

**Comment by dprus** [comment]
I'm still hitting this issue. @SimonPilkington, any suggestions?

**Comment by SimonPilkington** [comment]
Sorry about ignoring this. For some reason I got no notification when troglobit atted me.

According to the log, Cloudflare doesn't like what you're sending it (it's returning HTTP 400), but I can't reproduce this. The problem seems to be specific to your config, so can you post it?

**Comment by dprus** [comment]
Sure, here's my config, it's pretty basic really:

> period          = 300                                                                                                                                                
user-agent      = Mozilla/5.0                                                                                                                                        
allow-ipv6      = true                                                                                                                                               
                                                                                                                                                                     
provider cloudflare.com {                                                                                                                                            
     username = me@my.domain                                                                                                                                   
     password = FJbOphlZy0YrtV3p38GyGotRuKfhheO5sPUx3fJu                                                                                                             
     hostname = router.home.my.domain                                                                                                                               
}

I pasted in my real API Token above but then deleted it from Cloudflare directly after. So it's useless now.

**Comment by Lefey** [comment]
Hello! i have same problem.
Inadyn compiled from master branch.
Config:
`period          = 300
user-agent      = Mozilla/5.0

provider cloudflare.com {
     username = my-email@gmail.com/my-zone-name.com
     password = my-cloudflare-global-api-token
     hostname = router.my-zone-name.com
}
`
Inadyn log:
`inadyn[1]: In-a-dyn version 2.6-dev -- Dynamic DNS update client.
inadyn[1]: Guessing DDNS plugin 'default@cloudflare.com' from 'cloudflare.com'
inadyn[1]: Update forced for alias router.my-zone-name.com, new IP# 100.100.100.100
inadyn[1]: Cloudflare says our request was invalid.
inadyn[1]: Zone 'my-zone-name.com' not found.
`

**Comment by Lefey** [comment]
Sorry, i found my problem itself - missing ssl ca certificates, i compiled inadyn on other host system (first try was the router firmware) and cloudflare dns update work like a charm.
@dprus you can try to run inadyn with key --loglevel=debug to see you problem, it helped me.

**Comment by SimonPilkington** [comment]
@dprus, there is nothing conspicuous about your config. Lefey's suggestion is the best I can recommend you try. If that doesn't work for you, I may need to make a patch for to enable more verbose debugging for you so we can figure this out. I probably should have included request/response printing at debug verbosity to begin with.

**Comment by dprus** [comment]
I'm using a Docker container created using the Dockerfile from this github repository. It already installs ca-certificates:

```
RUN apk --no-cache add \
  ca-certificates \
  confuse \
  gnutls
```

Do I need more than that for Cloudflare to work?

**Comment by dprus** [comment]
That doesn't seem to be my problem. Looks like my certificates are fine. Here's the debug output:
```
inadyn    | <html><head><title>Current IP Check</title></head><body>Current IP Address: x.x.x.x</body></html>
inadyn    | inadyn[1]: Checking IPv4 address x.x.x.x ...
inadyn    | inadyn[1]: IPv4 address x.x.x.x is valid.
inadyn    | inadyn[1]: No IP# change detected for default@cloudflare.com, still at x.x.x.x
inadyn    | inadyn[1]: Update forced for alias router.home.my.domain, new IP# x.x.x.x
inadyn    | inadyn[1]: User: me@my.domain Zone: my.domain
inadyn    | inadyn[1]: Id query, initiating HTTPS ...
inadyn    | inadyn[1]: Id query, connecting to api.cloudflare.com([104.19.193.29]:443)
inadyn    | inadyn[1]: Certificate OK
inadyn    | inadyn[1]: SSL server cert subject: C=US,ST=CA,L=San Francisco,O=CloudFlare\, Inc.,CN=api.cloudflare.com
inadyn    | inadyn[1]: SSL server cert issuer: C=US,ST=CA,L=San Francisco,O=CloudFlare\, Inc.,CN=CloudFlare Inc ECC CA-2
inadyn    | inadyn[1]: Successfully sent HTTPS request!
inadyn    | inadyn[1]: Successfully received HTTPS response (1020 bytes)!
inadyn    | inadyn[1]: Cloudflare says our request was invalid.
inadyn    | inadyn[1]: Zone 'my.domain' not found.
```

**Comment by SimonPilkington** [comment]
All right, I'll put together a patch with more verbose debugging output this weekend probably.

**Comment by SimonPilkington** [comment]
Had some time just now so this is done already. @dprus please compile from my master branch and see what happens. Take care not to post any information you'd like to keep private.

@troglobit should we merge this?

**Comment by troglobit** [comment]
@SimonPilkington definitely, just open a PR :)

**Comment by dprus** [comment]
OK, here's the new output:

```
inadyn    | <html><head><title>Current IP Check</title></head><body>Current IP Address: x.x.x.x</body></html>                                                                                                                                                    
inadyn    | inadyn[1]: Checking IPv4 address x.x.x.x ...                                                                                                                                                                                                         
inadyn    | inadyn[1]: IPv4 address x.x.x.x is valid.                                                                                                                                                                                                            
inadyn    | inadyn[1]: No IP# change detected for default@cloudflare.com, still at x.x.x.x                                                                                                                                                                       
inadyn    | inadyn[1]: Update forced for alias router.home.my.domain, new IP# x.x.x.x                                                                                                                                                                           
inadyn    | inadyn[1]: User: me@my.domain Zone: my.domain                                                                                                                                                                                                      
inadyn    | inadyn[1]: Id query, initiating HTTPS ...                                                                                                                                                                                                                 
inadyn    | inadyn[1]: Id query, connecting to api.cloudflare.com([104.19.192.29]:443)                                                                                                                                                                                
inadyn    | inadyn[1]: Certificate OK                                                                                                                                                                                                                                 
inadyn    | inadyn[1]: SSL server cert subject: C=US,ST=CA,L=San Francisco,O=CloudFlare\, Inc.,CN=api.cloudflare.com                                                                                                                                                  
inadyn    | inadyn[1]: SSL server cert issuer: C=US,ST=CA,L=San Francisco,O=CloudFlare\, Inc.,CN=CloudFlare Inc ECC CA-2                                                                                                                                              
inadyn    | inadyn[1]: Request:                                                                                                                                                                                                                                       
inadyn    | GET /client/v4/zones?name=my.domain HTTP/1.0                                                                                                                                                                                                             
inadyn    | Host: api.cloudflare.com                                                                                                                                                                                                                                  
inadyn    | User-Agent: Mozilla/5.0                                                                                                                                                                                                                                   
inadyn    | Accept: */*                                                                                                                                                                                                                                               
inadyn    | X-Auth-Email: me@my.domain                                                                                                                                                                                                                          
inadyn    | X-Auth-Key: MY_SUPER_LONG_KEY_IS_HERE                                                                                                                                                                                                      
inadyn    | Content-Type: application/json                                                                                                                                                                                                                            
inadyn    |                                                                                                                                                                                                                                                           
inadyn    | inadyn[1]: Successfully sent HTTPS request!                                                                                                                                                                                                               
inadyn    | inadyn[1]: Successfully received HTTPS response (1020 bytes)!                                                                                                                                                                                             
inadyn    | inadyn[1]: Response:                                                                                                                                                                                                                                      
inadyn    | HTTP/1.1 400 Bad Request                                                                                                                                                                                                                                  
inadyn    | Date: Fri, 13 Dec 2019 23:02:18 GMT                                                                                                                                                                                                                       
inadyn    | Content-Type: application/json                                                                                                                                                                                                                            
inadyn    | Connection: close                                                                                                                                                                                                                                         
inadyn    | Set-Cookie: __cfduid=d1a42c12c36274774d366c69697f6368b1576278138; expires=Sun, 12-Jan-20 23:02:18 GMT; path=/; domain=.api.cloudflare.com; HttpOnly                                                                                                       
inadyn    | CF-Ray: 544b909efadfbb76-SEA                                                                                                                                                                                                                              
inadyn    | Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0                                                                                                                                                                             
inadyn    | Expires: Sun, 25 Jan 1981 05:00:00 GMT                                                                                                                                                                                                                    
inadyn    | Set-Cookie: vses2=e2sb7ump21jccfrkorng8sbmgtq6n0n6; path=/; domain=dash.cloudflare.com; secure; HttpOnly                                                                                                                                                  
inadyn    | Strict-Transport-Security: max-age=31536000, max-age=31536000                                                                                                                                                                                             
inadyn    | CF-Cache-Status: DYNAMIC                                                                                                                                                                                                                                  
inadyn    | Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"                                                                                                                                                        
inadyn    | Pragma: no-cache                                                                                                                                                                                                                                          
inadyn    | Served-In-Seconds: 0.016                                                                                                                                                                                                                                  
inadyn    | X-Content-Type-Options: nosniff                                                                                                                                                                                                                           
inadyn    | X-Frame-Options: SAMEORIGIN                                                                                                                                                                                                                               
inadyn    | Server: cloudflare                                                                                                                                                                                                                                        
inadyn    |                                                                                                                                                                                                                                                           
inadyn    | {"success":false,"errors":[{"code":6003,"message":"Invalid request headers","error_chain":[{"code":6103,"message":"Invalid format for X-Auth-Key he                                                                                                       
inadyn    | inadyn[1]: Cloudflare says our request was invalid.                                                                                                                                                                                                       
inadyn    | inadyn[1]: Zone 'my.domain' not found.     
```

**Comment by dprus** [comment]
I notice it says the key is in an invalid format. But I'm pasting it straight from Cloudflare's website so I'm not sure exactly what I'm doing wrong there?

**Comment by SimonPilkington** [comment]
Neither am I, frankly, but if I remove a single character from my API key, I get the exact error you're getting. Can you try fetching your zone ID using curl?

Use this command line:
```
curl -i -X GET https://api.cloudflare.com/client/v4/zones?name=YOUR_ZONE -H "X-Auth-Email: YOUR_EMAIL" -H "X-Auth-Key: YOUR_API_KEY"
```

**Comment by holy-systems-emperor** [comment]
Cloudflare released support for API tokens with granular permissions a couple months ago, which is what @dprus appears to be using. It's a good choice over using the Global API Key that plugin code currently supports, but the calling convention is a bit different for them. You can read about the release [here](https://blog.cloudflare.com/api-tokens-general-availability/):

tl;dr, while using the Global API Key requires applying the `X-Auth-Email` and `X-Auth-Key` headers, API Tokens only require `Authorization: Bearer <YOUR_API_KEY>`.

@SimonPilkington, how do you feel about removing handling of Global API Keys and switching over to using API Tokens exclusively? The Global API Key confers elevated privileges across the entire gamut of Cloudflare services, so it would be nice to give a dynamic DNS client an API Token that will only permit the bearer to update DNS records.

**Comment by SimonPilkington** [comment]
That definitely sounds like something we should do, yes. I'm actually still not using Cloudflare as I'm waiting for my plugin to make it to AsusWRT Merlin, but it looks like that's not happening before the 2.6 release, so I haven't heard about API tokens.

Thanks for letting me know. I'll look into it sometime soon.

**Comment by SimonPilkington** [comment]
#283 should solve this issue. @dprus, if you're still around please check if you're good to go now.

**Comment by dprus** [comment]
Sure. I will have to set things up again since I moved over to a different solution temporarily. Give me a few days and I will verify all is well. Thanks!

-Richard.

________________________________

**Comment by troglobit** [comment]
Hi guys, I'm preparing the v2.6 release this weekend.  Would be great to hear back from you, @dprus, if you managed to get some time to check the #283 fix.

**Comment by holy-systems-emperor** [comment]
On my end, I can attest that builds of #283 (4976bfe) and a more recent master (b44fa63) can successfully update Cloudflare's DNS records. The latter build also properly updates ipv6 records when the `allow-ipv6` option is enabled. Thanks @SimonPilkington, @jorhett, and @troglobit for taking the time to revise and flesh out this feature.

**Comment by troglobit** [comment]
Awesome feedback, @holy-systems-emperor, this is going to be a great release! :-)

**Comment by troglobit** [comment]
v2.6 released, finally ...

**Comment by troglobit** [comment]
@dprus I hope the move went smoothly. For purpose of completeness I'd love to hear back on this issue, at your earliest convenience of course :)

**Comment by dprus** [comment]
Yeah – I’ve moved over permanently to a different solution so I will have to setup inadyn again just temporarily to validate this. I still plan on doing so just haven’t had a chance yet. I’m hoping soon. Sorry for the delay…

-Richard.

**Comment by troglobit** [comment]
@dprus No problem, thank you for taking the time to respond :)

**Comment by jamesdmurphy** [comment]
This doesn't appear to be fixed however it could be a bug with the cloudflare API https://community.cloudflare.com/t/bug-zone-detail-by-name-requires-zone-list-permission/128042

**Comment by RichardVine** [comment]
Hi,

I've come across what seems to be a similar issue with a 'Zone not found' error which may be related. However, it doesn't seem to be a Cloudflare permissions issue but with the wrong domain being picked up.

I'm using inadyn 2.7 on an asuswrt-merlin router. My slightly redacted config file is:

provider cloudflare.com {
     checkip-command = "/usr/sbin/nvram get wan0_ipaddr"
     username = richard@exampledomain.co.uk
     password = *cloudflareapitoken*
     hostname = www.exampledomain.co.uk
}

Although I do get a Cloudflare permissions error, this is because the domain is detected as 'co.uk', which of course, I don't own. The correct zone should be exampledomain.co.uk. A non debug log shows:

inadyn[2177]: In-a-dyn version 2.7 -- Dynamic DNS update client.
inadyn[2177]: Guessing DDNS plugin 'default@cloudflare.com' from 'cloudflare.com'
inadyn[2177]: Update forced for alias www.exampledomain.co.uk, new IP# nnn.nnn.nnn.nnn
inadyn[2177]: HTTP 403: Provided API token does not have the required permissions.
inadyn[2177]: Zone 'co.uk' not found.
inadyn[2177]: Error response from DDNS server, exiting!


I tried specifying the username as 'richard@dummy.exampledomain.co.uk' in case the zone detection was just stripping the top level domain and using the rest, but this made no difference; co.uk was still detected.

I'm not a C programmer but within the cloudflare plugin, there seems to be a 'get_zones' function which returns up to the last two domain qualifiers, ie. co.uk in my case rather than the full domain. 

I don't know if the username parameter is used for anything else, but it appears that simply using  everything after the '@' would give the zone name. If that's not an option (because the username may not actually relate to the hostname being updated), using the hostname to determine the zone name could also be an issue, eg. some.server.exampledomain.co.uk would also likely detect the wrong zone. Perhaps a new 'zone' parameter is required.



Relevant part of inadyn debug log and response from Cloudflare:

inadyn[814]: User: richard@dummy.exampledomain.co.uk Zone: co.uk
...
...
inadyn[814]: Request:
GET /client/v4/zones?name=co.uk HTTP/1.0
Host: api.cloudflare.com
User-Agent: inadyn/2.7 https://github.com/troglobit/inadyn/issues
Accept: */*
Authorization: Bearer *cloudflareapitoken*
Content-Type: application/json
inadyn[814]: Successfully sent HTTPS request!
inadyn[814]: Successfully received HTTPS response (964 bytes)!
inadyn[814]: Response:
HTTP/1.1 403 Forbidden
Date: Wed, 06 May 2020 07:14:47 GMT
Content-Type: application/json
Connection: close
Set-Cookie: __cfduid=d97595a9bd6a69604d74db5e54d88a7281588749286; expires=Fri, 05-Jun-20 07:14:46 GMT; path=/; domain=.api.cloudflare.com; HttpOnly; SameSite=Lax
CF-Ray: 58f0e8034ca1f42f-LHR
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Expires: Sun, 25 Jan 1981 05:00:00 GMT
Strict-Transport-Security: max-age=31536000
CF-Cache-Status: DYNAMIC
Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
Pragma: no-cache
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
Server: cloudflare
cf-request-id: 028a6f56080000f42f48084200000001

{"success":false,"errors":[{"code":0,"message":"Actor 'com.cloudflare.api.token.39cc5833cbdb43c368429fad1d7ee0cb' requires permission 'com.cloudflare.api.account.zone.list' to list zones"}],"messages":[],"result":null}
inadyn[814]: HTTP 403: Provided API token does not have the required permissions.
inadyn[814]: Zone 'co.uk' not found.

**Comment by SimonPilkington** [comment]
Thanks for reporting @RichardVine; this is a different problem and not related to the original issue. Will be prusued at #312.

**Comment by troglobit** [comment]
@SimonPilkington thank you for moving this to #312, it is indeed a separate issue! :)

**Comment by kevin-leong** [comment]
Hello I am getting some what of a similar issue.  If you scroll to the bottom of my debug log, you will see "Failed to parse JSON."  I am not sure what is wrong here or why its failing to parse the JSON.

For the API token, I gave the permission of "Zone - Zone:Read" and "Zone - DNS:Edit".  I also omitted my IP, API Token Key, and username (which is set to my zone per the instructions), and hostname in the below log.

I appreciate if someone can please take a look at this.  Please feel free to let me know if you need any further information.

Thank you.

```
 # inadyn -f /config/inadyn.conf --loglevel=debug --once --foreground 
inadyn[81]: In-a-dyn version 2.7 -- Dynamic DNS update client.
inadyn[81]: Guessing DDNS plugin 'default@cloudflare.com' from 'cloudflare.com'
inadyn[81]: Resolving hostname [###OMIT###] => IP# 172.67.186.158
inadyn[81]: Base64 encoded string: c3luaWNhbC5vcmc6bnBobUtVbGlLOUxWWHk5dG4wdmp2SUZXSzY5OW5uU1o5YnI4UjhIdw==
inadyn[81]: Get address for default@cloudflare.com
inadyn[81]: Checking for IP# change, connecting to checkip.dyndns.com([131.186.113.70]:80)
inadyn[81]: Querying DDNS checkip server for my public IP#: GET / HTTP/1.0
Host: checkip.dyndns.com
User-Agent: Mozilla/5.0

inadyn[81]: Server response: HTTP/1.1 200 OK
Content-Type: text/html
Server: DynDNS-CheckIP/1.2.0
Connection: close
Cache-Control: no-cache
Pragma: no-cache
Content-Length: 104

<html><head><title>Current IP Check</title></head><body>Current IP Address: [###OMIT###]</body></html>
inadyn[81]: Checked my IP, return code 0: OK
inadyn[81]: IP server response:
inadyn[81]: HTTP/1.1 200 OK
Content-Type: text/html
Server: DynDNS-CheckIP/1.2.0
Connection: close
Cache-Control: no-cache
Pragma: no-cache
Content-Length: 104

<html><head><title>Current IP Check</title></head><body>Current IP Address: [###OMIT###]</body></html>
inadyn[81]: Checking IPv4 address [###OMIT###] ...
inadyn[81]: IPv4 address [###OMIT###] is valid.
inadyn[81]: Current IP# [###OMIT###] at default@cloudflare.com
inadyn[81]: Update forced for alias [###OMIT###], new IP# [###OMIT###]
inadyn[81]: User: [###OMIT###] Zone: [###OMIT###]
inadyn[81]: Id query, initiating HTTPS ...
inadyn[81]: Id query, connecting to api.cloudflare.com([104.19.192.29]:443)
inadyn[81]: Certificate OK
inadyn[81]: SSL server cert subject: C=US,ST=CA,L=San Francisco,O=Cloudflare\, Inc.,CN=api.cloudflare.com
inadyn[81]: SSL server cert issuer: C=US,O=Cloudflare\, Inc.,CN=Cloudflare Inc ECC CA-3
inadyn[81]: Request:
GET /client/v4/zones?name=[###OMIT###] HTTP/1.0
Host: api.cloudflare.com
User-Agent: Mozilla/5.0
Accept: */*
Authorization: Bearer [###OMIT###]
Content-Type: application/json

inadyn[81]: Successfully sent HTTPS request!
inadyn[81]: Successfully received HTTPS response (2738 bytes)!
inadyn[81]: Response:
HTTP/1.1 200 OK
Date: Sun, 16 Aug 2020 14:14:03 GMT
Content-Type: application/json
Connection: close
Set-Cookie: __cfduid=daeaaff8f8491621ec4af7fb3368cc1cd1597587243; expires=Tue, 15-Sep-20 14:14:03 GMT; path=/; domain=.api.cloudflare.com; HttpOnly; SameSite=Lax; Secure
CF-Ray: 5c3bc2709d74fce1-KIX
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Expires: Sun, 25 Jan 1981 05:00:00 GMT
Set-Cookie: __cflb=0H28vgHxwvgAQtjUGU4vq74ZFe3sNVUZYrhLadNDGjw; SameSite=Lax; path=/; expires=Sun, 16-Aug-20 16:44:04 GMT; HttpOnly
Strict-Transport-Security: max-age=31536000
CF-Cache-Status: DYNAMIC
cf-request-id: 049937da5d0000fce15aa4b200000001
Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
Pragma: no-cache
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
Server: cloudflare

{"result":[{"id":"da0eb4c1d5651d91124f0a0e98924646","name":"[###OMIT###]","status":"active","paused":f
inadyn[81]: Failed to parse JSON.
inadyn[81]: Zone '[###OMIT###]' not found.
```
